### DISCUSSION TITLE
ERR_REQUIRE_ESM from `esm.mjs`

### DISCUSSION THREAD

**Discussion by dyllan-to-you** [TypeStrong/ts-node#1820] [Q&A] [answered]
Hi!

Node 16.15.1
Typescript 4.7.4
ts-node 10.8.1

Whenever I try to run ts-node, utilizing ESM, I get the following error:
```
node:internal/modules/cjs/loader:979
    throw new ERR_REQUIRE_ESM(filename, true);
    ^

[Error [ERR_REQUIRE_ESM]: require() of ES Module C:\Users\DyllanTo\Desktop\Projects\analytics\algorithms\algorithms-js\node_modules\ts-node\esm.mjs not supported.
Instead change the require of C:\Users\DyllanTo\Desktop\Projects\analytics\algorithms\algorithms-js\node_modules\ts-node\esm.mjs to a dynamic import() which is available in all CommonJS modules.] {
  code: 'ERR_REQUIRE_ESM'
}
```

It looks like it's triggering on the following lines from `esm.mjs`
```
/** @type {import('./dist/esm')} */
const esm = require('./dist/esm');
```

I'm running the following command to call tsnode: `node -r ts-node/esm src/runner`. 

My tsconfig:
```json
{
  "ts-node": {
    "experimentalResolver": true
  },
  "compilerOptions": {
    "moduleResolution": "node",
    "module": "esnext",
    "target": "es6",
    "lib": ["ESNext", "DOM"],
    "allowJs": true,
    "outDir": "./dist",
    "sourceMap": true,
    "resolveJsonModule": true,
    "esModuleInterop": true,
    "strict": false,
    "strictNullChecks": false
  },
  "include": ["./src/**/*.ts", "./src/**/*.js"],
  "exclude": ["./src/**/*.test.ts", "./dist", ".eslintrc.cjs"]
}
```

I also have `type: module` in my package.json. 

I've no idea what I'm doing wrong, or if it's actually a bug
Any help would be greatly appreciated!

**Comment by cspotcode** [comment, marked as answer]
`node -r ts-node/esm` is the mistake.  Our docs explain correct usage: https://typestrong.org/ts-node/docs/imports

`-r ts-node/esm` tells node to require `ts-node/esm` which points to `node_modules/ts-node/esm.mjs`.  `esm.mjs` is an ESM module and cannot be required, which is why you get that error.  It was a fair guess to think `dist/esm` might be related, but in this case it is not.

Aside from that, `ts-node/esm` is a node ESM loader so must be used with `--loader`, not `-r`.
